wardrobe/kitties/src/lib.rs

```diff
@@ -544,18 +572,85 @@ impl SimpleConstraintChecker for FreeKittyConstraintChecker {
                 Ok(0)
             }
             Self::Breed => {
+                // Check that we are consuming at least one input
                 ensure!(input_data.len() == 2, Self::Error::TwoParentsDoNotExist);
+
                 let mom = KittyData::try_from(&input_data[0])?;
                 let dad = KittyData::try_from(&input_data[1])?;
                 KittyHelpers::can_breed(&mom, &dad)?;
-
                 // Output must be Mom, Dad, Child
                 ensure!(output_data.len() == 3, Self::Error::NotEnoughFamilyMembers);
-
                 KittyHelpers::check_new_family(&mom, &dad, output_data)?;
-
                 Ok(0)
             }
+            Self::UpdateKittyName => {
+                can_kitty_name_be_updated(input_data, output_data)?;
+                Ok(0)
+            }
+        }
+    }
+}
+
+/// Checks:
+///     - Input and output is of kittyType
+///     - Only name is updated and ther basicproperties are not updated.
+///
+pub fn can_kitty_name_be_updated(
+    input_data: &[DynamicallyTypedData],
+    output_data: &[DynamicallyTypedData],
+) -> Result<TransactionPriority, ConstraintCheckerError> {
+    ensure!(
+        input_data.len() == output_data.len() && !input_data.is_empty(),
+        { ConstraintCheckerError::InvalidNumberOfInputOutput }
+    );
+
+    let mut map: BTreeMap<KittyDNA, KittyData> = BTreeMap::new();
+
+    for utxo in input_data {
+        let utxo_kitty = utxo
+            .extract::<KittyData>()
+            .map_err(|_| ConstraintCheckerError::BadlyTyped)?;
+        map.insert(utxo_kitty.clone().dna, utxo_kitty);
+    }
+
+    for utxo in output_data {
+        let utxo_output_kitty = utxo
+            .extract::<KittyData>()
+            .map_err(|_| ConstraintCheckerError::BadlyTyped)?;
+
+        if let Some(input_kitty) = map.remove(&utxo_output_kitty.dna) {
+            // Element found, access the value
+            check_kitty_name_update(&input_kitty, &utxo_output_kitty)?;
+        } else {
+            return Err(ConstraintCheckerError::OutputUtxoMissingError);
         }
     }
+    return Ok(0);
+}
```

Comment:
Creating this data structure and searching through it are unnecessary on chain computation. If you want to allow multiple renames at all, then require the user to specify the inputs and outputs in the same order.

Reply:
@JoshOrndorff, Thanks for review.So If I understand correctly, to avoid more processing in the on-chain such as map(BTree) traversal, we don't want to do this right?I intended to reduce the number of transactions by clubbing all the renames in one which will reduce the gas cost if users want to and provide the user flexibility to send the input and output in any order. Then I will check the input and output just based on the index or I will support single rename.

Reply:
Supporting a single rename is fine.

Supporting multiple is also fine, but don't do the map. If the user wants to rename three kitties, they have to come in the same order in the inputs and the outputs.

Reply:
> I intended to reduce the number of transactions by clubbing all the renames in one

This is reasonable

> and provide the user flexibility to send the input and output in any order.

This is not.

Reply:
Thanks, @JoshOrndorff I will modify and resubmit the PR.
One follow-up question.
Map usage should be avoided in all other cases such as listKittyForSale,DelistKittyFromSale,updatePrice also right?